### PR TITLE
First iteration for AddNode

### PR DESCRIPTION
**Pull request by holisticode**
Closes https://ava-labs.atlassian.net/browse/FRAM-296

WARNING: Creating this PR resulted in an *unexpected* merge conflict in a *generated* gRPC file `rpcpb/rpc.pb.go` during the necessary `git rebase` operation to fix the merge conflict. 

This was pretty surprising. 

The script `./scripts/genproto.sh` was then run again which overwrote the conflicting generated file. It might have fixed the underlying issue, but this PR should be carefully reviewed.

Also take the time to consider what exactly should be provided to a new independent node (whitelisted subnets, plugin dir, etc.).

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -194,6 +195,22 @@ func (c *client) Stop(ctx context.Context) (*rpcpb.StopResponse, error) {
 	return c.controlc.Stop(ctx, &rpcpb.StopRequest{})
 }
 
+func (c *client) AddNode(ctx context.Context, name string, execPath string, opts ...OpOption) (*rpcpb.AddNodeResponse, error) {
+	ret := &Op{}
+	ret.applyOpts(opts)
+
```
Comment (gyuho) [inline client/client.go:201]: Can we follow the pattern we have in `RestartNode` so optional fields are just set to null?

```go
	if ret.whitelistedSubnets != "" {
		req.WhitelistedSubnets = &ret.whitelistedSubnets
	}
	if ret.logLevel != "" {
		req.LogLevel = &ret.logLevel
	}
	if ret.rootDataDir != "" {
		req.RootDataDir = &ret.rootDataDir
	}
```

File: server/server.go [hunk 2]
```diff
@@ -514,6 +515,64 @@ func (s *server) recvLoop(stream rpcpb.ControlService_StreamStatusServer) error
...
+	zap.L().Debug("received add node request", zap.String("name", req.Name))
+
+	if info := s.getClusterInfo(); info == nil {
+		return nil, ErrNotBootstrapped
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	nodeName := req.Name
```
Comment (gyuho) [inline server/server.go:528]: Can we error if the name is in conflict (e.g., `s.network.nodeInfos`)?

Can we error if the specified binary path does not exist?

File: client/client.go (hunk 1 above)
Comment (holisticode) [reply client/client.go:201]: done

File: server/server.go (hunk 2 above)
Comment (holisticode) [reply server/server.go:528]: I tested this and it fails anyway later when running `s.network.nw.AddNode`.
However it makes sense to check and fail early, so will change accordingly.

**Comment by danlaine** [comment]
@holisticode  This doesn't seem to work just yet.

In one tab, do:
```
avalanche-network-runner server \
--log-level debug \
--port=":8080" \
--grpc-gateway-port=":8081"
```

In another, do:
```
avalanche-network-runner control start \
--log-level debug \               
--endpoint="0.0.0.0:8080" \
--avalanchego-path ${AVALANCHEGO_EXEC_PATH}
dialing endpoint "0.0.0.0:8080"
```

and then:
```
avalanche-network-runner control add-node --log-level=info --avalanchego-path $AVALANCHEGO_EXEC_PATH --node-name bob
```

This causes the client to crash.

Also, can we please add an E2E test for this and document this new functionality in the README?

File: server/server.go [hunk 3]
```diff
@@ -514,6 +516,82 @@ func (s *server) recvLoop(stream rpcpb.ControlService_StreamStatusServer) error
...
+		return nil, fmt.Errorf("couldn't generate staking Cert/Key: %w", err)
+	}
+	nodeConfig := node.Config{
+		Name:               req.Name,
+		ConfigFile:         configFile,
+		ImplSpecificConfig: json.RawMessage(fmt.Sprintf(`{"binaryPath":"%s","redirectStdout":true,"redirectStderr":true}`, execPath)),
+		StakingKey:         string(stakingKey),
+		StakingCert:        string(stakingCert),
+	}
+	s.network.nodeNames = append(s.network.nodeNames, req.Name)
```
Comment (danlaine) [inline server/server.go:575]: I think we should do this after we've called `s.network.nw.AddNode` in case that errors

File: server/server.go [hunk 4]
```diff
@@ -514,6 +516,82 @@ func (s *server) recvLoop(stream rpcpb.ControlService_StreamStatusServer) error
...
+	if info := s.getClusterInfo(); info == nil {
+		return nil, ErrNotBootstrapped
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	var logLevel, whitelistedSubnets, pluginDir string
+
+	nodeName := req.Name
```
Comment (danlaine) [inline server/server.go:531]: I don't think we need this variable

**Comment by danlaine** [comment]
left 2 small comments, also bumping the above: `Also, can we please add an E2E test for this and document this new functionality in the README?`

File: server/server.go (hunk 4 above)
Comment (holisticode) [reply server/server.go:531]: This is a debatable nit really, but ok.

Comment (danlaine) [reply server/server.go:531]: We should use one or the other but not both

**Review by danlaine — APPROVED** [review]

**Pull request merged**